applications/StructuralMechanicsApplication/custom_strategies/custom_builder_and_solver/explicit_builder_and_solver.hpp

```diff
@@ -0,0 +1,453 @@
+        }
+        
+        KRATOS_CATCH( "" )
+    }
+
+    //**************************************************************************
+    //**************************************************************************
+
+    void BuildRHS(
+        typename TSchemeType::Pointer pScheme,
+        ModelPart& rModelPart,
+        TSystemVectorType& b)
+    {
+        KRATOS_TRY
+        
+        // Compute condition contributions to RHS.
+        CalculateAndAddConditionsRHS(pScheme, rModelPart);
+
+        // Compute element contributions to RHS.
+        CalculateAndAddElementsRHS(pScheme, rModelPart);
+
+        
+        KRATOS_CATCH( "" )
+
+    }
+
+
+    
+    
+    //***************************************************************************
+    //***************************************************************************
+
+   
+    void CalculateAndAddConditionsRHS(typename TSchemeType::Pointer pScheme, ModelPart& rModelPart )
+    {
+        KRATOS_TRY
+
+        ProcessInfo& rCurrentProcessInfo      = rModelPart.GetProcessInfo();
+        ConditionsArrayType& pConditions      = rModelPart.Conditions();
+
+        #ifdef _OPENMP
+            int number_of_threads = omp_get_max_threads();
+        #else
+            int number_of_threads = 1;
+        #endif
+
+        vector<unsigned int> condition_partition;
+        OpenMPUtils::CreatePartition(number_of_threads, pConditions.size(), condition_partition);
+
+
+        #pragma omp parallel for 
+        for(int k=0; k<number_of_threads; k++)
+        {
+        typename ConditionsArrayType::ptr_iterator it_begin=pConditions.ptr_begin()+condition_partition[k];
+        typename ConditionsArrayType::ptr_iterator it_end=pConditions.ptr_begin()+condition_partition[k+1];
+
+        for (typename ConditionsArrayType::ptr_iterator it= it_begin; it!=it_end; ++it)
+        {
+
+            LocalSystemVectorType RHS_Condition_Contribution = LocalSystemVectorType(0);
+
+            Element::EquationIdVectorType equation_id_vector_dummy; //Dummy
+
+            pScheme->Condition_Calculate_RHS_Contribution(*it, RHS_Condition_Contribution, equation_id_vector_dummy, rCurrentProcessInfo);
+        }
+        }
+
+        KRATOS_CATCH("")
+    }
+
+    
+    //***************************************************************************
+    //***************************************************************************
+
+
+    void CalculateAndAddElementsRHS(typename TSchemeType::Pointer pScheme, ModelPart& rModelPart )
+    {
+
+        KRATOS_TRY
+        
+        ProcessInfo& rCurrentProcessInfo    = rModelPart.GetProcessInfo();
+        ElementsArrayType& pElements        = rModelPart.Elements();
+
+        #ifdef _OPENMP
+            int number_of_threads = omp_get_max_threads();
+        #else
+            int number_of_threads = 1;
+        #endif
+
+        vector<unsigned int> element_partition;
+        OpenMPUtils::CreatePartition(number_of_threads, pElements.size(), element_partition);
```

Comment:
i would avoid doing the partitioning here

instead i think it is better to rely on openmp's partitioning capabilities

Reply:
fixed